src/effects/shaders/halftone.frag

```diff
@@ -0,0 +1,205 @@
+		dist_c_3 = distanceToDotRadius( c.samp3, c.p3, c.normal, p, angle, radius );
+		dist_c_4 = distanceToDotRadius( c.samp4, c.p4, c.normal, p, angle, radius );
+		res = ( dist_c_1 > 0.0 ) ? clamp( dist_c_1 / aa, 0.0, 1.0 ) : 0.0;
+		res += ( dist_c_2 > 0.0 ) ? clamp( dist_c_2 / aa, 0.0, 1.0 ) : 0.0;
+		res += ( dist_c_3 > 0.0 ) ? clamp( dist_c_3 / aa, 0.0, 1.0 ) : 0.0;
+		res += ( dist_c_4 > 0.0 ) ? clamp( dist_c_4 / aa, 0.0, 1.0 ) : 0.0;
+		res = clamp( res, 0.0, 1.0 );
+
+		return res;
+
+}
+
+Cell getReferenceCell( vec2 p, vec2 origin, float grid_angle, float step ) {
+
+		// get containing cell
+		Cell c;
+
+		// calc grid
+		vec2 n = vec2( cos( grid_angle ), sin( grid_angle ) );
+		float threshold = step * 0.5;
+		float dot_normal = n.x * ( p.x - origin.x ) + n.y * ( p.y - origin.y );
+		float dot_line = -n.y * ( p.x - origin.x ) + n.x * ( p.y - origin.y );
+		vec2 offset = vec2( n.x * dot_normal, n.y * dot_normal );
+		float offset_normal = mod( hypot( offset.x, offset.y ), step );
+		float normal_dir = ( dot_normal < 0.0 ) ? 1.0 : -1.0;
+		float normal_scale = ( ( offset_normal < threshold ) ? -offset_normal : step - offset_normal ) * normal_dir;
+		float offset_line = mod( hypot( ( p.x - offset.x ) - origin.x, ( p.y - offset.y ) - origin.y ), step );
+		float line_dir = ( dot_line < 0.0 ) ? 1.0 : -1.0;
+		float line_scale = ( ( offset_line < threshold ) ? -offset_line : step - offset_line ) * line_dir;
+
+		// get closest corner
+		c.normal = n;
+		c.p1.x = p.x - n.x * normal_scale + n.y * line_scale;
+		c.p1.y = p.y - n.y * normal_scale - n.x * line_scale;
+
+		// scatter
+		if ( scatter != 0.0 ) {
+
+			float off_mag = scatter * threshold * 0.5;
+			float off_angle = rand( vec2( floor( c.p1.x ), floor( c.p1.y ) ) ) * PI2;
+			c.p1.x += cos( off_angle ) * off_mag;
+			c.p1.y += sin( off_angle ) * off_mag;
+
+		}
+
+		// find corners
+		float normal_step = normal_dir * ( ( offset_normal < threshold ) ? step : -step );
+		float line_step = line_dir * ( ( offset_line < threshold ) ? step : -step );
+		c.p2.x = c.p1.x - n.x * normal_step;
+		c.p2.y = c.p1.y - n.y * normal_step;
+		c.p3.x = c.p1.x + n.y * line_step;
+		c.p3.y = c.p1.y - n.x * line_step;
+		c.p4.x = c.p1.x - n.x * normal_step + n.y * line_step;
+		c.p4.y = c.p1.y - n.y * normal_step - n.x * line_step;
+
+		return c;
+
+}
+
+vec4 mainImage(const in vec4 inputColor, const in vec2 uv, const in GData gData) {
+
+	// setup
+	vec2 p = vec2( uv.x * resolution.x, uv.y * resolution.y );
+	vec2 origin = vec2( 0, 0 );
+	float aa = ( radius < 2.5 ) ? radius * 0.5 : 1.25;
+
+	// get channel samples
+	Cell cell_r = getReferenceCell( p, origin, rotateR, radius );
+	Cell cell_g = getReferenceCell( p, origin, rotateG, radius );
+	Cell cell_b = getReferenceCell( p, origin, rotateB, radius );
+	float r = getDotColor( cell_r, p, 0, rotateR, aa );
+	float g = getDotColor( cell_g, p, 1, rotateG, aa );
+	float b = getDotColor( cell_b, p, 2, rotateB, aa );
+
+	return vec4( r, g, b, 1.0 );
+
+}
```

Comment:
There are some code style issues present in this shader, e.g. snake_case → camelCase, `( ` →`(` && ` )` →`)` and the indentation is off by one tab in some of the functions.